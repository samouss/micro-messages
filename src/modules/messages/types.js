// @flow

export type MessageId = string;
export type MessageBody = string;

export type Visibility = 'public' | 'private';

export type VisibilityFilter = Visibility | 'all';

export type Message = {
  id: MessageId,
  date: string,
  body: MessageBody,
  visibility: Visibility,
};

export type MessageInput = {
  body: MessageBody,
  visibility: Visibility,
};

export type ByIdState = { [id: MessageId]: Message };
export type IdsState = Array<MessageId>;

export type MessageState = {
  byId: ByIdState,
  ids: IdsState,
};

export type FetchMessagesRequestAction = { type: 'FETCH_MESSAGES_REQUEST' };
export type FetchMessagesSuccessAction = { type: 'FETCH_MESSAGES_SUCCESS', messages: Array<Message> };
export type FetchMessagesFailureAction = { type: 'FETCH_MESSAGES_FAILURE' };

export type PostMessageRequestAction = { type: 'POST_MESSAGE_REQUEST', message: Message };
export type PostMessageSuccessAction = { type: 'POST_MESSAGE_SUCCESS', message: Message };
export type PostMessageFailureAction = { type: 'POST_MESSAGE_FAILURE' };

export type MessageAction =
  | FetchMessagesRequestAction
  | FetchMessagesSuccessAction
  | FetchMessagesFailureAction

  | PostMessageRequestAction
  | PostMessageSuccessAction
  | PostMessageFailureAction;
