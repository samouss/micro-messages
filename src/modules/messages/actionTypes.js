// @flow

export const FETCH_MESSAGES_REQUEST = 'FETCH_MESSAGES_REQUEST';
export const FETCH_MESSAGES_SUCCESS = 'FETCH_MESSAGES_SUCCESS';
export const FETCH_MESSAGES_FAILURE = 'FETCH_MESSAGES_FAILURE';

export const POST_MESSAGE_REQUEST = 'POST_MESSAGE_REQUEST';
export const POST_MESSAGE_SUCCESS = 'POST_MESSAGE_SUCCESS';
export const POST_MESSAGE_FAILURE = 'POST_MESSAGE_FAILURE';
