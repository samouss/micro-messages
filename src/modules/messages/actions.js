// @flow

import { fetchMessages as fetchMessagesFromApi } from 'core/api';
import type { Dispatch } from 'store/types';
import * as ACTION_TYPES from './actionTypes';
import type { Message, MessageAction } from './types';

const fetchMessagesRequest = (): MessageAction => ({
  type: ACTION_TYPES.FETCH_MESSAGES_REQUEST,
});

const fetchMessagesSuccess = (messages: Array<Message>): MessageAction => ({
  type: ACTION_TYPES.FETCH_MESSAGES_SUCCESS,
  messages,
});

const fetchMessagesFailure = (): MessageAction => ({
  type: ACTION_TYPES.FETCH_MESSAGES_FAILURE,
});

export const fetchMessages = () => (dispatch: Dispatch) => {
  dispatch(fetchMessagesRequest());

  return fetchMessagesFromApi().then(res => {
    dispatch(fetchMessagesSuccess(res));
  }).catch(() => {
    dispatch(fetchMessagesFailure());
  });
};
