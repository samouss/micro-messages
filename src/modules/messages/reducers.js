// @flow

import { combineReducers } from 'redux';
import * as ACTION_TYPES from './actionTypes';
import type { State, Action } from 'store/types';
import type { MessageId, ByIdState, IdsState } from './types';

const byIdInitialState: ByIdState = {};
const byId = (state = byIdInitialState, action: Action) => {
  switch (action.type) {
    case ACTION_TYPES.FETCH_MESSAGES_SUCCESS: {
      return {
        ...state,
        ...action.messages.reduce((acc, message) => ({
          ...acc,
          [message.id]: message,
        }), {}),
      };
    }

    default: {
      return state;
    }
  }
};

const idsInitialState: IdsState = [];
const ids = (state = idsInitialState, action: Action) => {
  switch (action.type) {
    case ACTION_TYPES.FETCH_MESSAGES_SUCCESS: {
      return state.concat(action.messages.map(message => message.id));
    }

    default: {
      return state;
    }
  }
};

export default combineReducers({
  byId,
  ids,
});

const getState = (state: State) => state.messages;
const getMessageIds = (state: State) => getState(state).ids;

export const getMessage = (state: State, id: MessageId) => {
  return getState(state).byId[id];
};

export const getMessages = (state: State) => {
  return getMessageIds(state).map(id => getMessage(state, id));
};
