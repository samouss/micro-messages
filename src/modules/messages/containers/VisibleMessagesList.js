// @flow

import React, { Component } from 'react';
import { connect } from 'react-redux';
import Loader from 'components/Loader';
import { fetchMessages } from '../actions';
import { getVisibilityFilter, getMessagesWithVisibilityFilter } from '../reducers';
import MessagesList from '../components/MessagesList';

import type { State, Dispatch } from 'store/types';
import type { Message } from '../types';

type DefaultProps = {
  messages: Array<Message>,
};

export type Props = {
  messages: Array<Message>,
  fetchMessages: () => Promise<void>,
};

type ComponentState = {
  isLoading: boolean,
};

export class VisibleMessagesList extends Component {
  static defaultProps: DefaultProps;
  props: Props;
  state: ComponentState;

  constructor(props: Props) {
    super(props);

    this.state = {
      isLoading: true,
    };
  }

  componentDidMount() {
    this.fetchMessages();
  }

  fetchMessages() {
    return this.props.fetchMessages().then(() => {
      this.setState({ isLoading: false });
    }).catch(() => {
      this.setState({ isLoading: false });
    });
  }

  render() {
    if (this.state.isLoading) {
      return <Loader />;
    }

    return (
      <MessagesList
        messages={this.props.messages}
      />
    );
  }
}

VisibleMessagesList.defaultProps = {
  messages: [],
};

const mapStateToProps = (state: State) => ({
  messages: getMessagesWithVisibilityFilter(state, getVisibilityFilter(state)),
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  fetchMessages: () => dispatch(fetchMessages()),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(VisibleMessagesList);
