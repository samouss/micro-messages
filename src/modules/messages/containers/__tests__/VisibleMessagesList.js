// @flow

import React from 'react';
import { shallow } from 'enzyme';
import { createMockMessage } from 'test/messages';
import { VisibleMessagesList } from '../VisibleMessagesList';

import type { Props } from '../VisibleMessagesList';

describe('<VisibleMessagesList />', () => {
  const defaultProps = {
    messages: [
      createMockMessage('ID1234'),
      createMockMessage('ID5678'),
      createMockMessage('ID9101'),
    ],
    fetchMessages: () => Promise.resolve(),
  };

  it('expect to render the component correctly with Loader on first render', () => {
    const props: Props = {
      ...defaultProps,
    };

    const component = shallow(
      <VisibleMessagesList
        {...props}
      />,
    );

    expect(component).toMatchSnapshot();
  });

  it('expect to render the component correctly with MessagesList', () => {
    const props: Props = {
      ...defaultProps,
    };

    const component = shallow(
      <VisibleMessagesList
        {...props}
      />,
    );

    // $FlowFixMe
    component.setState({ isLoading: false });

    expect(component).toMatchSnapshot();
  });

  it('expect to call fetchMessages on didMount', () => {
    const props: Props = {
      ...defaultProps,
      fetchMessages: jest.fn(() => Promise.resolve()),
    };

    const component = shallow(
      <VisibleMessagesList
        {...props}
      />,
    );

    // $FlowFixMe
    component.instance().componentDidMount();

    expect(props.fetchMessages).toHaveBeenCalled();
  });

  it('expect to call fetchMessages and setState', () => {
    expect.assertions(3);

    const props: Props = {
      ...defaultProps,
      fetchMessages: jest.fn(() => Promise.resolve()),
    };

    const expectation = {
      before: { isLoading: true },
      after: { isLoading: false },
    };

    const component = shallow(
      <VisibleMessagesList
        {...props}
      />,
    );

    // $FlowFixMe
    expect(component.state()).toEqual(expectation.before);

    // $FlowFixMe
    return component.instance().fetchMessages().then(() => {
      expect(props.fetchMessages).toHaveBeenCalled();
      // $FlowFixMe
      expect(component.state()).toEqual(expectation.after);
    });
  });
});
