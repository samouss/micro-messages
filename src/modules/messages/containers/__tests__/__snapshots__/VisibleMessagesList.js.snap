// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VisibleMessagesList /> expect to render the component correctly with Loader on first render 1`] = `<Loader />`;

exports[`<VisibleMessagesList /> expect to render the component correctly with MessagesList 1`] = `
<MessagesList
  messages={
    Array [
      Object {
        "body": "BODY",
        "date": "DATE",
        "id": "ID1234",
        "visibility": "private",
      },
      Object {
        "body": "BODY",
        "date": "DATE",
        "id": "ID5678",
        "visibility": "private",
      },
      Object {
        "body": "BODY",
        "date": "DATE",
        "id": "ID9101",
        "visibility": "private",
      },
    ]
  }
/>
`;
